### ISSUE TITLE
USB serial port loses data on 3.12.* kernels

### ISSUE THREAD

**Issue by EDkan** [raspberrypi/linux#725]
a couple of weeks ago, for some reasons, i upgraded my raspberry pi (rev.B) on newest "stable" kernel 3.12.28+ (by _rpi-update_) and then have some strange problems with communicating through USB-serial ports. 

my software uses modbus to query some devices linked with raspberry via number of USB-serial converters and have internal statistics, which says that on kernel 3.10.38+ (and earlier 3.6.11) there are no any errors(broken request-responses) for few weeks and longer.

in addition, when i set kernel flag dwc_otg.speed=1 then CPU load become 85% and error count reduced for many times but they are still there.

kernel **3.10.38+**, after about 24 hours:

| baudrate/byte timeout* | successes | errors | loss | CPU load |
| --- | --- | --- | --- | --- |
| 19200 / 5200 mks | 1630310 | 0 (!) | 0 | ~80% |
| 230400 / 2550 mks | 4818298 | 0 (!) | 0 | ~80% |

kernel **3.12.32+**, after 25 minutes:

| baudrate/byte timeout* | successes | errors | loss | CPU load |
| --- | --- | --- | --- | --- |
| 19200 / 5200 mks | 15802 | 49 | 0.310 | ~99% |
| 230400 / 2550 mks | 37634 | 18 | 0.048 | ~99% |

kernel **3.12.32+ with dwc_otg.speed=1**, after 3 hours:

| baudrate/byte timeout* | successes | errors | loss | CPU load |
| --- | --- | --- | --- | --- |
| 19200 / 5200 mks | 220982 | 3 | 0.0014 | ~85% |
| 230400 / 2550 mks | 704881 | 4 | 0.00057 | ~85% |

_*byte timeout_ is the maximum delay between two sequential bytes in packet; if it occur, received packet is considered complete

lsmod output:

```
Module                  Size  Used by
leds_gpio               2079  0
led_class               4118  1 leds_gpio
ftdi_sio               30672  7
usbserial              26684  15 ftdi_sio
```

lsusb:

```
Bus 001 Device 002: ID 0424:9514 Standard Microsystems Corp.
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 001 Device 003: ID 0424:ec00 Standard Microsystems Corp.
Bus 001 Device 004: ID 05e3:0610 Genesys Logic, Inc. 4-port hub
Bus 001 Device 005: ID 05e3:0610 Genesys Logic, Inc. 4-port hub
Bus 001 Device 006: ID 0403:6001 Future Technology Devices International, Ltd FT232 USB-Serial (UART) IC
Bus 001 Device 007: ID 0403:6001 Future Technology Devices International, Ltd FT232 USB-Serial (UART) IC
Bus 001 Device 008: ID 0403:6001 Future Technology Devices International, Ltd FT232 USB-Serial (UART) IC
```

and lsusb -v -d 0403:

```
Bus 001 Device 006: ID 0403:6001 Future Technology Devices International, Ltd FT232 USB-Serial (UART) IC
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 (Defined at Interface level)
  bDeviceSubClass         0
  bDeviceProtocol         0
  bMaxPacketSize0         8
  idVendor           0x0403 Future Technology Devices International, Ltd
  idProduct          0x6001 FT232 USB-Serial (UART) IC
  bcdDevice            6.00
  iManufacturer           1
  iProduct                2
  iSerial                 3
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength           32
    bNumInterfaces          1
    bConfigurationValue     1
    iConfiguration          0
    bmAttributes         0x80
      (Bus Powered)
    MaxPower              200mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass       255 Vendor Specific Class
      bInterfaceSubClass    255 Vendor Specific Subclass
      bInterfaceProtocol    255 Vendor Specific Protocol
      iInterface              2
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x02  EP 2 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               0
```

what additional tweaks i should do? what information i must present to help you to fix this issue?

and rhetorical:
what happened in 3.12. kernels that these errors appeared?
why 3.12.? May be other branches are more stable?

**Update:**
forgot to mention, that on other USB-serial devices such as, for example,

```
Bus 001 Device 004: ID 067b:2303 Prolific Technology, Inc. PL2303 Serial Port
```

there is the same issue. hence the problem possibly in `usbserial` module or deeper.

**Comment by P33M** [comment]
Is your error condition discovered at the protocol layer?

**Comment by EDkan** [comment]
At protocol layer i simply get timeout error, that caused i suppose by sending broken request to external device (that ignored it)

I wanted to gather some evidence by simple reproducible program or shell script, but the results are not clear. maybe later I will post here the code of my simple diagnostic program

here are results of simple program that acts like ping for serial line: 2 converters of same raspberry device connected by 3 line RX1 <--> TX2, TX1 <--> RX2, GND <--> GND. on second endpoint runs this command:
 `stty` _baudrate_ `raw -echo < /dev/ttyUSB1; cat < /dev/ttyUSB1 > /dev/ttyUSB1`
and on first endpoint - test program just sends data-string then receives it and compare results:

kernel **3.10.38+**, packet size 56 bytes, each test 300 seconds long:

| baudrate | byte timeout | packets sent | errors | loss, % | CPU load, % |
| --- | --- | --- | --- | --- | --- |
| 9600 | 34376 | 2912 | 0 | 0 | 10 |
| 19200 | 17188 | 5612 | 0 | 0 | 13 |
| 38400 | 8594 | 10604 | 2 | 0,02 | 15 |
| 57600 | 5730 | 15206 | 4 | 0,03 | 21 |
| 115200 | 2865 | 25892 | 12 | 0,05 | 29 |
| 230400 | 1433 | 38831 | 7561 | 19,47 | 34 |

kernel **3.12.32+**, packet size 56 bytes, each test 300 seconds long:

| baudrate | byte timeout | packets sent | errors | loss, % | CPU load, % |
| --- | --- | --- | --- | --- | --- |
| 9600 | 34376 | 3059 | 6(0) | 0,20(0) | 15 |
| 19200 | 17188 | 5861 | 7 | 0,12 | 16 |
| 38400 | 8594 | 10957 | 4 | 0,86 | 20 |
| 57600 | 5730 | 15946 | 281 | 1,76 | 24 |
| 115200 | 2865 | 26200 | 347 | 1,32 | 26 |
| 230400 | 1433 | 40510 | 115 | 0,28 | 32 |

test program code listing:

``` C++
#include <unistd.h>
#include <stdint.h>

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <fcntl.h>
#include <string.h>
#include <assert.h>

#include <termios.h>
#include <algorithm>

#include <time.h>

termios old_tios;

/* Sets up a serial port for RTU communications */
int connect_serial(const char* dev_name, int baudrate, int* portd)
{
    termios tios;
    speed_t speed;
    int flags;

    /* The O_NOCTTY flag tells UNIX that this program doesn't want
       to be the "controlling terminal" for that port. If you
       don't specify this then any input (such as keyboard abort
       signals and so forth) will affect your process

       Timeouts are ignored in canonical input mode or when the
       NDELAY option is set on the file via open or fcntl */
    flags = O_RDWR | O_NOCTTY | O_NDELAY | O_EXCL;
#ifdef O_CLOEXEC
    flags |= O_CLOEXEC;
#endif

    *portd = open(dev_name, flags);
    if (*portd == -1) {
            fprintf(stderr, "ERROR Can't open the device %s (%s)\r\n",
                    dev_name, strerror(errno));
        return -1;
    }

    /* Save */
    tcgetattr(*portd, &old_tios);

    memset(&tios, 0, sizeof(struct termios));

    /* C_ISPEED     Input baud (new interface)
       C_OSPEED     Output baud (new interface)
    */
    switch (baudrate) {
    case 110: speed = B110; break;
    case 300: speed = B300; break;
    case 600: speed = B600; break;
    case 1200: speed = B1200; break;
    case 2400: speed = B2400; break;
    case 4800: speed = B4800; break;
    case 9600: speed = B9600; break;
    case 19200: speed = B19200; break;
    case 38400: speed = B38400; break;
#ifdef B57600
    case 57600: speed = B57600; break;
#endif
#ifdef B115200
    case 115200: speed = B115200; break;
#endif
#ifdef B230400
    case 230400: speed = B230400; break;
#endif
    default:
        speed = B9600;
        fprintf(stderr, "WARNING Unknown baud rate %d for %s (B9600 used)\r\n",
                baudrate, dev_name);
    }

    /* Set the baud rate */
    if ((cfsetispeed(&tios, speed) < 0) ||
        (cfsetospeed(&tios, speed) < 0)) {
        close(*portd);
        *portd = -1;
        return -1;
    }

    /* C_CFLAG      Control options
       CLOCAL       Local line - do not change "owner" of port
       CREAD        Enable receiver
    */
    tios.c_cflag |= (CREAD | CLOCAL);
    /* CSIZE, HUPCL, CRTSCTS (hardware flow control) */

    /* CSIZE        Bit mask for data bits */
    tios.c_cflag &= ~CSIZE;
    tios.c_cflag |= CS8;


    /* Stop bit (1 or 2) */
    tios.c_cflag &=~ CSTOPB;
    ///tios.c_cflag |= CSTOPB;

    /* PARENB       Enable parity bit
       PARODD       Use odd parity instead of even */
    // no parity
    tios.c_iflag &= ~INPCK;
    tios.c_cflag &=~ PARENB;

    /* Raw input */
    tios.c_lflag &= ~(ICANON | ECHO | ECHOE | ISIG);

    /* Software flow control is disabled */
    tios.c_iflag &= ~(IXON | IXOFF | IXANY);

    /* Raw ouput */
    tios.c_oflag &= ~OPOST;

    /* Unused because we use open with the NDELAY option */
    tios.c_cc[VMIN] = 0;
    tios.c_cc[VTIME] = 0;

    if (tcsetattr(*portd, TCSAFLUSH/*TCSANOW*/, &tios) < 0) {
        close(*portd);
        *portd = -1;
        return -1;
    }

    return 0;
}

int disconnect_serial(int* portd) {
    if (*portd != -1) {
        tcsetattr(*portd, TCSANOW, &old_tios);
        close(*portd);
        *portd = -1;
    }
    return 0;
}

int _flush(int* portd) {
    return tcflush(*portd, TCIOFLUSH);
}

int _select(int* portd, fd_set *rset, struct timeval *tv)
{
    int s_rc;
    while ((s_rc = select(*portd+1, rset, NULL, NULL, tv)) == -1) {
        if (errno == EINTR) {
            fprintf(stderr, "A non blocked signal was caught\n");
            /* Necessary after an error */
            FD_ZERO(rset);
            FD_SET(*portd, rset);
        } else {
            return -1;
        }
    }

    return s_rc;
}

const char* saved_devname;
int saved_baudrate;

int _read(int* portd, uint8_t* in_buff, size_t in_buff_size, timeval byte_to)
{
    fd_set rset;
    FD_ZERO(&rset);
    FD_SET(*portd, &rset);
    timeval tv = {1, 0};
    int rc;
    int received = 0;

    while (_select(portd, &rset, &tv) > 0)
    {
        if (received >= in_buff_size) {
            _flush(portd);
            errno = ENOBUFS;
            return received;
        }

        tv.tv_sec = byte_to.tv_sec;
        tv.tv_usec = byte_to.tv_usec;
        rc = read(*portd, in_buff + received,
                std::min((size_t)8, in_buff_size - received));

        if (rc == 0) {
            errno = ECONNRESET;
            rc = -1;
        }

        if (rc == -1) {
            fprintf(stderr, "ERROR while read, try reconnect: %s\r\n", strerror(errno));
            if (errno == ECONNRESET || errno == ECONNREFUSED ||
                 errno == EBADF) {
                int saved_errno = errno;
                disconnect_serial(portd);
                connect_serial(saved_devname, saved_baudrate, portd);
                /* Could be removed by previous calls */
                errno = saved_errno;
            }
            return -1;
        }

        received += rc;
    }

    errno = 0;
    return received;
}

int _write(int* portd, const uint8_t* out_buff, size_t out_buff_size)
{
    return write(*portd, out_buff, out_buff_size);
}


int main(int argc, char** argv)
{
    saved_devname = "/dev/ttyUSBGen1";
    saved_baudrate = 19200;//9600;

    const char* test_string = "qwertyuiop[]asdfghjkl;'zxcvbnm,./1234567890-=+++++++++++";
    const size_t max_in_buff_size = 256;
    uint8_t in_buff[max_in_buff_size] = {0};

    int serial;
    if (connect_serial(saved_devname, saved_baudrate, &serial) < 0) {
        fprintf(stderr, "Error open serial port\r\n");
        return -1;
    }

    int byte_timeout_mks = std::max( 
        (10/*bits*/ * 1000000.0/*microseconds*/ / saved_baudrate) * 10/*bytes - minimum pause*/ + 1
        , 2500.0 );
    int sent = 0;
    int read = 0;
    timeval byte_to = {0, byte_timeout_mks};
    uint err_cnt = 0;
    time_t start = time(0);
    int time_to_run_s = 300;    // 5 minutes
    int write_time_mks = (10 * 1000000 / saved_baudrate) * strlen(test_string);
    printf("calculated byte-timeout %d, write-time (mks) %d\r\n", byte_timeout_mks, write_time_mks);

    int a = 0;
    for (; a < 100000; a++)
    {
        sent = _write(&serial, (uint8_t*)test_string, strlen(test_string));
        ///tcflush(serial, TCOFLUSH);

        //usleep(write_time_mks);
        bzero(in_buff, max_in_buff_size);
        read = _read(&serial, in_buff, max_in_buff_size, byte_to);
        (a % 10) == 0 && fprintf(stderr, "request %d: sent %d, read back %d\r", a, sent, read);
        if (strcmp(test_string, (char*)in_buff) != 0)
        {
            _flush(&serial);
            err_cnt++;
            fprintf(stderr, "\r\ncurrent err_count %d, data: %s\r\n", err_cnt, (char*)in_buff);
        }
        if (time(0) - start > time_to_run_s)
            break;
    }
    fprintf(stderr, "\r\niterations: %d, errors %d, runtime %d\r\n", a, err_cnt, time(0) - start);

    disconnect_serial(&serial);
    return 0;
}
```

**Comment by P33M** [comment]
Have you tested if the results are different if the test program is run on two Pis simultaneously?

(saves me having to unnecessarily replicate results)

**Comment by P33M** [comment]
With the test code on two USB-serial adapters (FT232R) plugged into an x86_64 kernel 3.2 machine, I get no errored transmissions.

With a single loopback USB-serial adapter plugged into a Pi or the PC, I get no errored transmissions.

I can see errors in approximately the quantities you describe when I use two USB-serial adapters plugged into the same Pi. This Pi is running default Raspbian (latest release). Errors also occur if I transplant the ttyUSB1 device to the PC.

**Comment by EDkan** [comment]
hello, @P33M!
thank you for being interested in this issue and trying to solve it.

> Have you tested if the results are different if the test program is run on two Pis simultaneously?

i haven't tested this program onto 2 PIs simultaneously, but i noticed same issues with 3.12. kernel on other PI (if you meant problems in broken hardware)

i'll now try some more tests (with two PIs ) and other...

> I can see errors in approximately the quantities you describe when I use two USB-serial adapters plugged into the same Pi. This Pi is running default Raspbian (latest release).

what kernel version uses in that latest release? is it 3.12.32?

could  you please downgrade kernel of your raspbian to 3.10.38+ ( `rpi-update f6eef32dd6388c3b04dbf462bd324d93281bf397` ) and run this test with 2 serial adapters in same PI? is there any differences in error counts and CPU usage?

**Comment by EDkan** [comment]
yesterday i little changed my test program code and made a series of tests with single USB-serial adapter loopbacked (screw fixed TX <--> RX contacts) and plugged into RPI. 

``` C++
    . . .

    int byte_timeout_mks = std::max( 
        (10/*bits*/ * 1000000.0/*microseconds*/ / saved_baudrate) * 10/*bytes - minimum pause*/ + 1
        , 2000.0 );
    int sent = 0;
    int read = 0;
    timeval byte_to = {0, byte_timeout_mks};
    uint err_cnt = 0;
    time_t start = time(0);
    int time_to_run_s = 300;    // 5 minutes
    int write_time_mks = (10 * 1000000 / saved_baudrate) * strlen(test_string);
    printf("calculated byte-timeout %d, write-time (mks) %d\r\n", byte_timeout_mks, write_time_mks);

    int a = 0;
    for (; a < 100000; a++)
    {
        sent = _write(&serial, (uint8_t*)test_string, strlen(test_string));
        ///tcflush(serial, TCOFLUSH);

        ///usleep(write_time_mks);
        bzero(in_buff, max_in_buff_size);
        read = _read(&serial, in_buff, max_in_buff_size, byte_to);
        (a % 50) == 0 && fprintf(stderr, "request %d: sent %d, read back %d\r", a, sent, read);
        if (strcmp(test_string, (char*)in_buff) != 0)
        {
            err_cnt++;
            fprintf(stderr, "request %d: sent %d, read back %d\r", a, sent, read);
            fprintf(stderr, "\r\ncurrent err_count %d, data: %s\r\n", err_cnt, (char*)in_buff);
        }
        if (time(0) - start > time_to_run_s)
            break;
    }
    fprintf(stderr, "\r\niterations: %d, errors %d, loss %% %0.2f, runtime %d\r\n", a, err_cnt, err_cnt * 100.0 / a, time(0) - start);

    disconnect_serial(&serial);
    return 0;
```

here are results (each pass is 300 seconds long):
kernel **3.12.32+**, with **dwc_otg.speed=1**

| baudrate | byte timeout | packets sent | errors | loss, % | CPU load, % |
| --- | --- | --- | --- | --- | --- |
| 9600 | 10417 | 4299 | 70 | 1,63 | 4 |
| 19200 | 5209 | 8527 | 45 | 0,53 | 6 |
| 38400 | 2605 | 16505 | 17 | 0,10 | 7 |
| 57600 | 2000 | 22967 | 7 | 0,03 | 7 |
| 115200 | 2000 | 37213 | 11 | 0,03 | 10 |
| 230400 | 2000 | 58771 | 6 | 0,01 | 14 |

kernel **3.12.32+**, **without dwc_otg.speed=1**

| baudrate | byte timeout | packets sent | errors | loss, % | CPU load, % |
| --- | --- | --- | --- | --- | --- |
| 9600 | 10417 | 4298 | 147 | 3,42 | 9 |
| 19200 | 5209 | 8631 | 796 | 9,22 | 11 |
| 38400 | 2605 | 16133 | 875 | 5,31 | 12 |
| 57600 | 2000 | 24038 | 836 | 3,48 | 14 |
| 115200 | 2000 | 39338 | 68 | 0,17 | 15 |
| 230400 | 2000 | 58510 | 58 | 0,10 | 17 |

kernel **3.10.38+**, dwc_otg.speed has no meaning

| baudrate | byte timeout | packets sent | errors | loss, % | CPU load, % |
| --- | --- | --- | --- | --- | --- |
| 9600 | 10417 | 4276 | 0 | 0 | 4 |
| 19200 | 5209 | 8316 | 0 | 0 | 4 |
| 38400 | 2605 | 16522 | 0 | 0 | 5 |
| 57600 | 2000 | 22493 | 0 | 0 | 6 |
| 115200 | 2000 | 36933 | 0 | 0 | 8 |
| 230400 | 2000 | 49460 | 0 | 0 | 10 |

**UPDATE (2014.11.21):**

kernel **3.12.32+**, **dwc_otg.fiq_fsm_enable=0**

| baudrate | byte timeout | packets sent | errors | loss, % | CPU load, % |
| --- | --- | --- | --- | --- | --- |
| 9600 | 10417 | 4310 | 262 | 6,08 | 8 |
| 19200 | 5209 | 8628 | 928 | 10,76 | 10 |
| 38400 | 2605 | 16165 | 989 | 6,12 | 11 |
| 57600 | 2000 | 23972 | 908 | 3,79 | 12 |
| 115200 | 2000 | 39240 | 50 | 0,13 | 12 |
| 230400 | 2000 | 58446 | 20 | 0,03 | 14 |

**Comment by P33M** [comment]
On a 3.12 kernel, what happens to the error rate if you add `dwc_otg.fiq_fsm_enable=0` to /boot/cmdline.txt?

**Comment by EDkan** [comment]
with `dwc_otg.fiq_fsm_enable=0` (and `without dwc_otg.speed`) option the error count and CPU load seems to be reduced a little (on higher baudrates) but they still remain. (btw, on lower baudrates count of errors increased). see table in update section of my previous comment.

**Comment by P33M** [comment]
An error rate of several % makes it much easier to view on an analyzer.

I will look at this when I next get a chance to.

**Comment by sndnvaps** [comment]
now i have a bug. 
in win7 & linux i cann't connect pidora with usb-ser.
pi kernel 3.12.*

**Comment by cmorty** [comment]
@EDkan Do you still have this problem? Do you get messages in `dmesg` about input overruns (just to confirm it is the same issue)? 

 I'm basically forwarding serial output to the network. I noticed two things: 
- Monitoring the network with `bmon`, the network seems to sometime hick-up, too, when loosing serial data. The transfer rate drops, but recovers without errors.
- running `ssh pi@raspi "cat /dev/ttyUSB0" > /dev/null` for both serials in parallel works fine. (I configured the tty appropriately before catting (using screen).)
  Especially the latter makes me wonder, as I am unable to trigger the issue even when creating a high network load.

**Comment by Ruffio** [comment]
@popcornmix consider closing this issue, as there has been no response to question asked to @EDkan 28 Apr 2015

**Comment by Ruffio** [comment]
Please close this issue, as there haven't been any activity since 28 Apr 2015, the Xcross issue in rpi-firmware was closed on 25 Nov 2014 and it is regarding 3.12.* kernel versions...
